runtime/src/chpl-gpu.c

```diff
@@ -827,6 +856,8 @@ static void launch_kernel(const char* name,
 
   cfg_finalize_priv_table(cfg);
 
+  cfg_set_halt_flag(cfg, 0);
```

Comment:
Currently, this is racy.

Each Chapel task has its own GPU stream. Multiple tasks can use the same GPU relatively safely thanks to that. If both of them tries to launch a kernel, those launches will be serialized by the driver. If one does communication, the other kernel, it'll overlap.

While the driver can serialize the launches, note that there will be a single `halt_flag` global in the generated device binary. Two Chapel tasks can race for that variable. e.g. one task could launch a kernel, which would set the flag, but before you can read it on the host, another task could set it back to 0.

I believe there is a way we can tell the device driver to synchronize streams in such a way that it recognizes setting/reading this flag is the same as kernel launch and should be part of the launch w.r.t. synchornization, but I don't know offhand the concoction for that.

What we have right now, instead, is to use our local spinlocks to synchronize these variables. We do that locking for `chpl_privateObjects` today. Check out `cfg_finalize_priv_table` and how `priv_table_lock` is acquired and released. I'd recommend having a similar lock for halt, or generalizing `priv_table_lock` to cover both cases.

As far as I understand, this is still an incomplete implementation (?), so that race is acceptable at this stage, but should be fixed before we call `halt` production-ready.

ps. I know that we discussed a race condition on this flag, but that was mostly on multiple gpu threads within the same kernel. I hadn't considered this aspect of it. I should have.